File: fsharp-asp.net/BookApi/BookApi/Controllers/BookController.fs
namespace BookApi.Controllers

open Microsoft.AspNetCore.Mvc
open BookApi
open System

[<ApiController>]
[<Route("api/[controller]")>]
type BookController (bookRepository : BookRepository) =
    inherit ControllerBase()

    [<HttpGet>]
    member _.Get() =
        bookRepository.Get()

    [<HttpGet("{id}")>]
    member _.GetById(id: Guid) : IActionResult = 
        match bookRepository.GetById(id) with
        | Some book -> new ObjectResult(book) :> IActionResult 
        | None -> NotFoundResult() :> IActionResult 


    [<HttpPost>]
    member _.Post(book: Book) : IActionResult =
        let addedBook = bookRepository.Add(book)
        CreatedAtRouteResult( "/" + addedBook.Id.ToString(), addedBook) :> IActionResult 

    [<HttpDelete("{id}")>]
    member _.Delete(id: Guid) : IActionResult = 
        bookRepository.Delete(id)
        NoContentResult() :> IActionResult

    [<HttpPut("{id}")>]
    member _.Update(id: Guid, book: Book) : IActionResult  =
        bookRepository.Update(id, book)
        NoContentResult() :> IActionResult

File: csharp-asp.net/BookApi/Controllers/BookController.cs
using BookApi.Models;
using BookApi.Repository;
using Microsoft.AspNetCore.Http.Extensions;
using Microsoft.AspNetCore.Mvc;
using System;

namespace BookApi.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class BookController : ControllerBase
    {
        private readonly BookRepository _bookRepository;

        public BookController(BookRepository bookRepository)
        {
            _bookRepository = bookRepository;
        }

        [HttpGet]
        public IActionResult Get()
        {
            var books = _bookRepository.Get();
            return Ok(books);
        }

        [HttpGet("{id}")]
        public IActionResult GetById([FromRoute] Guid id)
        {
            var book = _bookRepository.GetById(id);
            if (book == null)
            {
                return NotFound();
            }
            return Ok(book);
        }

        [HttpPost]
        public IActionResult Post([FromBody] Book book)
        {
            var addedBook = _bookRepository.Add(book);
            return Created(new Uri(Request.GetEncodedUrl() + "/" + addedBook.Id), addedBook);
        }

        [HttpPut("{id}")]
        public IActionResult Update([FromRoute] Guid id,[FromBody] Book book)
        {
            var updatedBook = _bookRepository.Update(id, book);
            if(updatedBook == null)
            {
                return NotFound();
            }
            return Ok();
        }

        [HttpDelete("{id}")]
        public IActionResult Delete([FromRoute] Guid id)
        {
            _bookRepository.Delete(id);
            return NoContent();
        }
    }
}
